lib/handler/cache-handler.js

```diff
@@ -0,0 +1,376 @@
+    super(handler)
+
+    this.#opts = opts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @param {Record<string, string | string[]> | undefined} headers
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    if (
+      this.#req.method in UNSAFE_METHODS &&
+      statusCode >= 200 &&
+      statusCode <= 399
+    ) {
+      // https://www.rfc-editor.org/rfc/rfc9111.html#name-invalidating-stored-respons
+      const result = this.#opts.store.deleteByOrigin(this.#req.origin)
+      if (
+        result &&
+        typeof result.catch === 'function' &&
+        typeof this.#handler.onError === 'function'
+      ) {
+        result.catch(this.#handler.onError)
+      }
+
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlHeader = headers['cache-control']
+    const contentLengthHeader = headers['content-length']
+
+    if (!cacheControlHeader || !contentLengthHeader) {
+      // Don't have the headers we need, can't cache
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const contentLength = Number(headers['content-length'])
+    if (
+      !Number.isInteger(contentLength) ||
+      this.#opts.store.isFull ||
+      this.#opts.store.entryCount >= this.#opts.store.maxEntries
+    ) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+    if (!canCacheResponse(statusCode, headers, cacheControlDirectives)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#writeStream = this.#opts.store.createWriteStream(this.#req, {
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        vary: varyDirectives,
+        cachedAt: now,
+        staleAt: now + staleAt,
+        deleteAt: now + deleteAt
```

Comment:
staleAt and deletAt means different things? one is time and other is duration? confusing

Reply:
`staleAt` is the time when the cached response is to be considered stale, but sometimes (like with `stale-while-revalidate`) we can serve a stale response up until a point in time, which is what `deleteAt` is. If we can't serve the response stale, they will equal each other.

Reply:
No I mean, `staleAt = now + staleAt` on the left side `staleAt` is an absolute time and on the right side it is a duration (?).

Reply:
Ahh fair,, will update so their determining functions return absolute time